### DISCUSSION TITLE
flatpak version for linux

### DISCUSSION THREAD

**Discussion by ryleu** [alvr-org/ALVR#1095] [General]
my quick pitch:
 - no build issues
 - no dealing with BRs for old versions
 - auto updates
 - better system integration
 - easier installing 

a flatpak version on flathub would have all of those benefits and more

**Comment by zmerp** [comment]
This was brought up multiple times, there are some insurmountable obstables, namely encapsulation. ALVR needs to modify SteamVR installation files in order to work.

**Reply by Promootheus** [reply]
Hi, does this mean that ALVR from the aur wouldn't work with a flatpak version of steam client for the same reason? Do both have to be installed as non-flatpak?

**Reply by ryleu** [reply]
@zarik5 the encapsulation can be relaxed. there are (fairly simple) ways to grant access to the correct directories to flatpak apps. you can literally just give access to `~/.steam/steam/steamapps/common/SteamVR`.

**Reply by zmerp** [reply]
I'm not a linux user and I go by what I hear in the Linux channels in the Discord server. Maybe @m00nwtchr can give you a more definitive answer.

**Reply by ryleu** [reply]
fair enough, i'll wait for their response.

**Reply by xytovl** [reply]
I don't know the details of flatpak sandbox, but the points that may cause issues are:
* ALVR needs to replace "vrcompositor" with a wrapper in order to force a few environment variables
* The environment variables reference a path to a vulkan layer, which will be loaded by vrcompositor (part of steam), but shipped by ALVR.
* ALVR needs to register itself in steamVR by adding an entry in "~/.config/openvr/openvr.vrpath", the entry will reference alvr's .so file. SteamVR then will need to load this shared library, and be able to find all the dependencies.
* The vulkan layer and alvr streaming server need to communicate, currently done through a pipe in "$XDG_RUNTIME_DIR", which must be visible by both steam and alvr.

Those are the main points that break sandboxing, if anyone knows flatpak well enough and can detail how to overcome this, it would help potential developers.

**Reply by jmaris** [reply]
Hey, perhaps I can contribute at least concerning two points: [GE-Proton](https://github.com/GloriousEggroll/proton-ge-custom) ( a version of proton built for gamers on linux ) is available as a flatpak and modifies the existing Steam flatpak (by adding it's custom proton lib)

Perhaps there will be some pointers there on gaining access to the steam flatpak sandbox.

If this is possible then it could address issues with ALVR registering itself, and with the pipe for communications.

**Reply by ryleu** [reply]
While the idea seems good, all the ProtonUp Flatpak does (which is what I believe you're referring to) is modifies the contents of Steam's library directory.

What could be useful, however, are the IPC permissions that you can request for a Flatpak. I believe that would let you talk to the Steam process with that. If not, there may be some other permission, or even an xdg portal, that you could use.

Edit: just refreshed myself on this thread. It turns out that one of the "insurmountable" issues is modifying the installation files of SteamVR, which is entirely possible in the constrains of a Flatpak.

**Reply by zmerp** [reply]
As a sidenote, a ALVR appimage has been successfully made to work https://github.com/alvr-org/ALVR/pull/1181

**Reply by jmaris** [reply]
> While the idea seems good, all the ProtonUp Flatpak does (which is what I believe you're referring to) is modifies the contents of Steam's library directory.
> 
> What could be useful, however, are the IPC permissions that you can request for a Flatpak. I believe that would let you talk to the Steam process with that. If not, there may be some other permission, or even an xdg portal, that you could use.
> 
> Edit: just refreshed myself on this thread. It turns out that one of the "insurmountable" issues is modifying the installation files of SteamVR, which is entirely possible in the constrains of a Flatpak.

There is also a possibility of using a location accessible to both for the pipe, but i'm not 100% sure how that would work in practice.

**Reply by himekifee** [reply]
Would it be possible to manually override some config to make it work then? Like manually editing the openvr runtime path and then passtheough the alvr so lib filesystem to the inner steam flatpak. I can think of some dl loading issue with different versions of other libraries. Could that be fixed by static linking? Other than that I think it's possible to achieve the goal. I need the steam flatpak because my vrchat does not launch on my arch native steam runtime due to some anti cheat runtime issues.
